File: yoonyoung/6주차/베스트_앨범.swift
//
//  main.swift
//  0926
//
//  Created by 조윤영 on 2021/09/26.
//

import Foundation

var answer: [Int] = []
func solution(_ genres:[String], _ plays:[Int]) -> [Int] {
    var map: [String: Int] = [:]
    for (idx, genre) in genres.enumerated() {
        if map[genre] == nil { map[genre] = plays[idx] }
        else { map[genre]! += plays[idx] }
    }

    var info: [(genre: String, sum: Int, playCnt: Int, idx: Int)] = []
    for (index, genre) in genres.enumerated() {
        info.append((genre: genre, sum: map[genre]!, playCnt: plays[index], idx: index))
    }

    info.sort(by: {
        if $0.sum == $1.sum {
            if $0.playCnt == $1.playCnt {
                return $0.idx < $1.idx
            } else { return $0.playCnt > $1.playCnt }
        }
        return $0.sum > $1.sum
    })

    var cnt = 0
    var prevGenre = ""
    for i in info {
        if prevGenre != i.genre {
            prevGenre = i.genre
            cnt = 1
        } else { cnt += 1 }

        if cnt > 2 { continue }
        answer.append(i.idx)
    }

    return answer
}

let genres = ["classic", "pop", "classic", "classic", "pop"]
let plays = [500, 600, 150, 800, 2500]
print(solution(genres, plays))

File: leoo/hash/베스트앨범.py
"""
https://programmers.co.kr/learn/courses/30/lessons/42579

- 속한 노래가 많이 재생된 장르를 먼저 수록합니다.
- 장르 내에서 많이 재생된 노래를 먼저 수록합니다.
- 장르 내에서 재생 횟수가 같은 노래 중에서는 고유 번호가 낮은 노래를 먼저 수록합니다.
"""
from collections import defaultdict


def get_sorted_genre(gen_itr, reverse=True):
    return sorted(gen_itr, key=lambda _: _[1], reverse=reverse)


def solution(genres, plays):
    gen_2_list = defaultdict(list)  # genre -> List[(index, play)]
    gen_2_counter = defaultdict(int)  # genre -> #_play

    for i, (g, p) in enumerate(zip(genres, plays)):
        gen_2_list[g].append((i, p))
        gen_2_counter[g] += p

    answer = []
    for g, count in get_sorted_genre(gen_2_counter.items()):
        for i, play in get_sorted_genre(gen_2_list[g])[:2]:  # 재생된 노래를 두 개씩 모아 베스트 앨범을 출시하려 합니다.
            answer.append(i)
    return answer


g = ["classic", "pop", "classic", "classic", "pop"]
p = [500, 600, 150, 800, 2500]
print(solution(g, p))
